server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -75,6 +82,22 @@ public RefreshListeners(IntSupplier getMaxRefreshListeners, Runnable forceRefres
         this.threadContext = threadContext;
     }
 
+    /**
+     * Prohibit adding new refresh listeners. See {@link #preventNewRefreshListeners}.
+     */
+    public void disallowAdd() {
+        synchronized (this) {
+            preventNewRefreshListeners = true;
+        }
+    }
+
+    /**
+     * Enable adding new refresh listeners. See {@link #preventNewRefreshListeners}.
+     */
+    public void allowAdd() {
+        preventNewRefreshListeners = false;
```

Comment:
synchronizing on this method is ok, it should be rarely called

Reply:
Done